### PR TITLE
 Ajax Call [AuthorizeForScopes] Attributed Action Produce CORS error when MSAL exception is thrown

### PR DESCRIPTION
**Pull request by creativebrother**
Controller Action decorated with [AuthorizeForScopes] Attribute called by ajax method cause user agent CORS error when MSAL exception is thrown.

Suggest enhancement
(1) in [AuthorizeForScopes] to provide a return url extract from ajax custom header for user agent to return to after the challenge is satisfied.
(2) in OpenIdConnentHandler,  after the redirect url for Identity Provider is calculated, for ajax request, the handler will return a 401 with this redirect url in the header location.

Added a test project to test modified AuthorizeForScopes Attribute to handle Ajax call. It depends on a locally built microsoft.aspnetcore.authentication.openidconnect dev nupkg within which associate change on ajax call redirection is handled differently (OpenIdConnectHandler need to be updated at Asp.net Core side).

### CODE REVIEW

**Comment by jennyf19** [comment]
@creativebrother thanks for the PR. we will take a look. Could you move the test project under the "tests" folder? thanks.

**Comment by creativebrother** [comment]
> @creativebrother thanks for the PR. we will take a look. Could you move the test project under the "tests" folder? thanks.

Hi, @jennyf19 I believe the test project is under tests folder. Do you mean I do not need an extra folder underneath tests folder, just put the test project directly underneath the tests folder?

**Comment by jennyf19** [comment]
@creativebrother like `tests\AjaxCallActionsWithDynamicConsent` or whatever you'd like to call the subfolder under `tests`. looking at the commit in GitHub (I haven't pulled your fork yet, but will), it looks like it's in the root of the project. thanks again.

**Comment by jennyf19** [comment]
thanks @creativebrother

**Comment by creativebrother** [comment]
@jennyf19 I made another PR in Aspnetcore project side. #26689

**Comment by jennyf19** [comment]
@creativebrother thanks again. I pulled your fork and tested it out. looks great. there are some things we would need to change, just small things on our end. I think it's easiest to merge your PR, and then we can update the test app as needed. this is really great. thank you. @jmprieur

---
In reply to: [705536463](https://github.com/AzureAD/microsoft-identity-web/pull/665#issuecomment-705536463) [](ancestors = 705536463)

File: tests/AjaxCallActionsWithDynamicConsent/appsettings.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+    "AzureAd": {
+        "Instance": "https://login.microsoftonline.com/",
+        "Domain": "xxx.onmicrosoft.com",
+        "TenantId": "xxx",
+        "ClientId": "xxx",
+        "ClientSecret": "xxx",
+        "CallbackPath": "/signin-oidc"
```
Comment (jennyf19) [inline tests/AjaxCallActionsWithDynamicConsent/appsettings.json:8]: cool. thanks for removing all this. was going to ask :)

File: tests/AjaxCallActionsWithDynamicConsent/Controllers/HomeController.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            _tokenAcquisition = tokenAcquisition;
+        }
+
+        public IActionResult Index()
+        {
+            return View();
+        }
+
+        [AuthorizeForScopes(Scopes = new[] { "https://ccbcc.sharepoint.com/AllSites.Read" })]
+        public async Task<IActionResult> AjaxAction()
```
Comment (jennyf19) [inline tests/AjaxCallActionsWithDynamicConsent/Controllers/HomeController.cs:30]: nit: if you want, you can put string.empty here or "user.read" as that is what we'll use to call graph. we can also update this later after it's merged.

File: tests/AjaxCallActionsWithDynamicConsent/Controllers/HomeController.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            return View(new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
+        }
+
+        /// <summary>
+        /// Private: Gets and returns an access token for the provided resource.
+        /// </summary>
+        /// <param name="resource">Resource to obtain access token for</param>
+        /// <returns></returns>
+        private async Task<string> GetAccessTokenforResource(string scope)
+        {
```
Comment (jennyf19) [inline tests/AjaxCallActionsWithDynamicConsent/Controllers/HomeController.cs:53]: nit: we'll probably remove this, but after we merge.

File: tests/AjaxCallActionsWithDynamicConsent/Views/Home/AjaxAction.cshtml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+﻿
+
+@{
+    Layout = null;
+}
+
+<h1>Controller Action decorated with [AuthorizeForScopes] Attribute called by regular and ajax method succesfully!</h1>
```
Comment (jennyf19) [inline tests/AjaxCallActionsWithDynamicConsent/Views/Home/AjaxAction.cshtml:7]: >succesfully [](start = 102, length = 11)

nit: successfully

File: tests/AjaxCallActionsWithDynamicConsent/Views/Home/AjaxAction.cshtml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    When user consent is needed, the above attribute will throw exception to challenge user to authenticate which is a redirect to Identity Provider,
+    if the framework does not emit a 401, ajax call will be "transparently" redirect to the Identity Provider and causing CORS.
+</p>
+<p>
+    Developer can add logic in OnRedirectToIdentityProvider to avoid that but it is not trivia for the average developer to figure out the cause of
+    CORS error initially and will spend time to investigate the issue themselves, and they will need to duplicate the portion of the code from framework
+    to obtain the redirect url anyway.
+</p>
+<p>
+    Fix: Similar as CookieAuthenticationHandler, The framework(both OpenIdConnectHandler and AuthorizeForScopes Attribute) check if the challenge is for
```
Comment (jennyf19) [inline tests/AjaxCallActionsWithDynamicConsent/Views/Home/AjaxAction.cshtml:19]: nit: space between "framework(.."

File: src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs [hunk 6]
```diff
@@ -99,6 +100,14 @@ public override void OnException(ExceptionContext context)
                         msalUiRequiredException,
                         context.HttpContext.User,
                         UserFlow);
+
+                    if (IsAjaxRequest(context.HttpContext.Request) && (!string.IsNullOrEmpty(context.HttpContext.Request.Headers["x-ReturnUrl"])
+                        || !string.IsNullOrEmpty(context.HttpContext.Request.Query["x-ReturnUrl"])))
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs:105]: >x-ReturnUrl [](start = 84, length = 11)

nit: x-ReturnUrl can go into our contants file (also something we can do post-merge). :)

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Comment by jennyf19** [comment]
#603 & #642

**Comment by jennyf19** [comment]
@creativebrother going to merge this and then i'll do another PR fixing the comments i have above. Thanks again for this AWESOME contribution. The team really appreciates it!

File: tests/AjaxCallActionsWithDynamicConsent/appsettings.json (hunk 1 above)
Comment (creativebrother) [reply tests/AjaxCallActionsWithDynamicConsent/appsettings.json:8]: Reminded myself but forgot about it. I have deleted the App registration on Azure anyway.

File: tests/AjaxCallActionsWithDynamicConsent/Controllers/HomeController.cs (hunk 2 above)
Comment (creativebrother) [reply tests/AjaxCallActionsWithDynamicConsent/Controllers/HomeController.cs:30]: Thanks.

File: tests/AjaxCallActionsWithDynamicConsent/Controllers/HomeController.cs (hunk 3 above)
Comment (creativebrother) [reply tests/AjaxCallActionsWithDynamicConsent/Controllers/HomeController.cs:53]: Yes, I agree. I was using an older version of our app as template and it is in the resource concept era....

File: src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs (hunk 6 above)
Comment (creativebrother) [reply src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs:105]: Thanks, it should be.

**Comment by creativebrother** [comment]
> @creativebrother going to merge this and then i'll do another PR fixing the comments i have above. Thanks again for this AWESOME contribution. The team really appreciates it!

@jennyf19 Love to contribute to Open Source Project. Your guys/ladies are awesome!

**Comment by jennyf19** [comment]
@creativebrother we love it to! less work for us! lol! idk when we would have gotten around to this. feel free to contribute as much as you like ;)

**Comment by OnAzureCloud9** [comment]
Thanks for the addition @creativebrother. I am very happy to see this incoming to 1.2.0 and I'm sure many others in the future will appreciate it as well.

**Pull request merged**